### DISCUSSION TITLE
Valid Ipopt options for CppAD

### DISCUSSION THREAD

**Discussion by type-null** [coin-or/CppAD#166] [Q&A]
When tinkering with [Ipopt options](https://coin-or.github.io/Ipopt/OPTIONS.html#:~:text=%E2%96%B8-,acceptable_obj_change_tol,-%3A%20%22Acceptance%22%20stopping%20criterion), I tried to adjust "acceptable_obj_change_tol" to terminate the program when the objective didn't improve much. 

```markdown
`acceptable_obj_change_tol`: "Acceptance" stopping criterion based on objective function change.

If the relative change of the objective function (scaled by Max(1,|f(x)|)) is less than this value, this part of the acceptable tolerance termination is satisfied; see also acceptable_tol. This is useful for the quasi-Newton option, which has trouble to bring down the dual infeasibility. The valid range for this real option is 0 ≤ acceptable_obj_change_tol and its default value is 10+20.
```

However, no matter set to 1e5000 or 1e-5000, with`acceptable_iter` set to 1, the iteration number of my program didn't change.

Then, I tried to append a nonsense option and it didn't give error message.

I wonder if `acceptable_obj_change_tol` and possibly other options are supported by CppAD or not. If so, how should I use this threshold correctly to stop the program earlier.

Many thanks!

**Comment by bradbell** [comment]
Are you using the interface described under options on ?
https://cppad.readthedocs.io/en/latest/ipopt_solve.html#options

**Reply by type-null** [reply]
Yep, here is how I used options in my code:

```c++
// object that computes objective and constraints
FG_eval fg_eval(m);

// options
string options;
// printing level: 0-12, default 5
options += "Integer print_level  5\n";
//linear solver choosing: m27, m86, m77, m97
options += "String linear_solver   ma97\n";
//derivative testing: none, first-order, second-order, only-second-order
options += "String derivative_test   none\n";

options += "Numeric acceptable_constr_viol_tol   1e-7\n";

options += "Numeric acceptable_obj_change_tol       1e50000\n";

options += "Numeric acceptable_iter       1\n";

options += "Numeric hahaha     1\n";


CppAD::ipopt::solve_result<Dvector> newSolution; // solution
CppAD::ipopt::solve<Dvector, FG_eval>(options, x0, xl, xu, gl, gu, fg_eval, newSolution); // solve the problem
```

**Comment by bradbell** [comment]
1. Try adding:   options += "Integer print_level  5\n";
2. What is your setting for tol (the convergence tolerance) ?

**Reply by type-null** [reply]
1. That's my first options as in my last reply
2. I set it to `options += "Numeric tol          1e-6\n";`

How does `acceptable_obj_change_tol` work? Since its default value is 10^20, every iteration should have satisfied this criterion. I also set `acceptable_iter` to 1 so why not my program stop at 2 iterations but continues as it likes?

**Comment by bradbell** [comment]
I have not used accceptable_obj_change_tol or acceptable_iter.
 I have been using tol and max_iter instead; see the heading Termination Criteria on
https://coin-or.github.io/Ipopt/OPTIONS.html#OPT_Termination

What does your optimizer trace look like when you set the print_level to 5 ?

**Reply by type-null** [reply]
Like this:
```cmd
# of nonzero in Jacobin, hessian,...

iterations

exit message
```

**Comment by bradbell** [comment]
It should look like this:
his is Ipopt version 3.13.4, running with linear solver mumps.
NOTE: Other linear solvers might be more efficient (see Ipopt documentation).

Number of nonzeros in equality constraint Jacobian...:        0
Number of nonzeros in inequality constraint Jacobian.:        0
Number of nonzeros in Lagrangian Hessian.............:        0

Total number of variables............................:      121
                     variables with only lower bounds:        0
                variables with lower and upper bounds:      121
                     variables with only upper bounds:        0
Total number of equality constraints.................:        0
Total number of inequality constraints...............:        0
        inequality constraints with only lower bounds:        0
   inequality constraints with lower and upper bounds:        0
        inequality constraints with only upper bounds:        0

iter    objective    inf_pr   inf_du lg(mu)  ||d||  lg(rg) alpha_du alpha_pr  ls
   0  1.3338628e-01 0.00e+00 1.00e+00   0.0 0.00e+00    -  0.00e+00 0.00e+00   0
   1  1.6012478e+00 0.00e+00 4.77e+00  -1.1 7.45e-02    -  9.98e-01 1.00e+00f  1
   2  1.6503250e+00 0.00e+00 1.65e+00  -1.2 2.88e-01    -  1.00e+00 3.44e-01f  1
   3  2.0201332e+00 0.00e+00 4.54e+00  -1.2 2.17e-01    -  1.00e+00 1.00e+00f  1
 ...

I am interested in the inf_pr and inf_du values.

**Comment by bradbell** [comment]
perhaps it would be best if you e-mail me an entire trace at bradbell@seanet.com
